lib/spack/spack/spec.py

```diff
@@ -5071,6 +5225,41 @@ def do_parse(self):
 
         try:
             while self.next:
+                if self.accept(SINGLE_QUOTE):
```

Comment:
I've always found this section of the code really difficult to follow because of all the nested if/else's.

I'd suggest any new additions should be functions so we get a better description of what is going on through the name. Maybe something like:

```python
def starts_with_quotes(self, quote_type, specs):
   # same code for single and double

def process_initial_quotes(self, specs):
   # replace both new if statements with this function
   starts_with_quotes(SINGLE_QUOTE, specs)
   starts_with_quotes(DOUBLE_QUOTE, specs)
```